include/benchmark/minimal_benchmark.h

```diff
@@ -0,0 +1,487 @@
+  //    BENCHMARK(Foo)->ThreadRange(1,16);
+  // This will run the following benchmarks:
+  //    Foo in 1 thread
+  //    Foo in 2 threads
+  //    Foo in 4 threads
+  //    Foo in 8 threads
+  //    Foo in 16 threads
+  MinimalBenchmark& ThreadRange(int min_threads, int max_threads);
+
+  // Equivalent to ThreadRange(NumCPUs(), NumCPUs())
+  MinimalBenchmark& ThreadPerCpu();
+
+  MinimalBenchmark* operator->() {
+    return this;
+  }
+  operator Benchmark*() {
+    Benchmark *tmp = imp_;
+    imp_ = nullptr;
+    return tmp;
+  }
+
+private:
+  Benchmark *imp_;
+  BENCHMARK_DISALLOW_COPY_AND_ASSIGN(MinimalBenchmark);
+};
+
+
+}  // end namespace benchmark
+
+// ------------------------------------------------------
+// Macro to register benchmarks
+
+// Helpers for generating unique variable names
+#define BENCHMARK_CONCAT(a, b, c) BENCHMARK_CONCAT2(a, b, c)
+#define BENCHMARK_CONCAT2(a, b, c) a ## b ## c
+
+#define BENCHMARK(n)                                                    \
+  static ::benchmark::Benchmark*                                          \
+  BENCHMARK_CONCAT(_benchmark_, n, __LINE__) BENCHMARK_UNUSED =        \
+  (::benchmark::MinimalBenchmark(#n, &n))
+
+// Old-style macros
+#define BENCHMARK_WITH_ARG(n, a) BENCHMARK(n)->Arg((a))
+#define BENCHMARK_WITH_ARG2(n, a1, a2) BENCHMARK(n)->ArgPair((a1), (a2))
+#define BENCHMARK_RANGE(n, lo, hi) BENCHMARK(n)->Range((lo), (hi))
+#define BENCHMARK_RANGE2(n, l1, h1, l2, h2) \
+  BENCHMARK(n)->RangePair((l1), (h1), (l2), (h2))
+
+// This will register a benchmark for a templatized function.  For example:
+//
+// template<int arg>
+// void BM_Foo(int iters);
+//
+// BENCHMARK_TEMPLATE(BM_Foo, 1);
+//
+// will register BM_Foo<1> as a benchmark.
+// NOTE: In C++11 BENCHMARK_TEMPLATE is variadic and will accept 1 or more
+// template arguments.
+#if __cplusplus < 201103L
+# define BENCHMARK_TEMPLATE(n, a)                                           \
+    static ::benchmark::Benchmark*                                          \
+    BENCHMARK_CONCAT(_benchmark_, n, __LINE__) BENCHMARK_UNUSED =           \
+    (::benchmark::MinimalBenchmark(#n "<" #a ">", &n<a>))
+#else
+# define BENCHMARK_TEMPLATE(n, ...)                                           \
+    static ::benchmark::Benchmark*                                            \
+    BENCHMARK_CONCAT(_benchmark_, n, __LINE__) BENCHMARK_UNUSED =             \
+    (::benchmark::MinimalBenchmark(#n "<" #__VA_ARGS__ ">", &n<__VA_ARGS__>))
+#endif
+
+#define BENCHMARK_TEMPLATE2(n, a, b)                                      \
+  static ::benchmark::Benchmark*                                          \
+  BENCHMARK_CONCAT(__benchmark_, n, __LINE__) BENCHMARK_UNUSED =          \
+  (::benchmark::MinimalBenchmark(#n "<" #a "," #b ">",                    \
+                               &n<a, b>))
+
+// Helper macro to create a main routine in a test that runs the benchmarks
+#define BENCHMARK_MAIN()                             \
```

Comment:
This works just fine for my cases. However I would like to build benchmark as a shared library and I don't think you can have a benchmark_main as a shared library.

We could provide both?